text/0122-sdk-hub-scope-merge.md

```diff
@@ -0,0 +1,148 @@
+
+// get the currently active client. May return a NOOP client. Replacement for `getCurrentHub().getClient()`.
+export function getClient(): Client;
+
+// make a scope the current scope. Replacement for `makeMain(hub)`
+export function makeCurrentScope(scope: Scope): void;
+```
+
+The following APIs already exist but will behave differently:
+
+* `withScope()` will still work, but it will actually fork an execution context. So this will roughyl the same as doing `hub.run()` today.
+
+APIs that are currently on the hub should instead be called directly on the scope (e.g. `scope.captureException()` etc.), or via a global method (e.g. `Sentry.captureException()`).
+
+## Clients
+
+Instead of a client being optional, there will now _always_ be a client. It may be a Noop Client that does nothing, if `init()` has not been called yet.
+
+Each client gets a fresh initial scope when it is initialized:
+
+```js
+const client = new Client();
+client.getScope(); // <-- fresh, initial scope for this client
+scope.getClient(); // <-- get client of scope
+```
+
+You can make a scope the current one like this:
+
+```js
+makeCurrentScope(client.getScope());
+// this scope/client will now be available via e.g. `getScope()` or `getClient()`
+// this is what init() calls under the hood
+```
+
+The current scope may be kept similar to how we currently keep the current hub, but this is SDK specific and not part of this RFC.
+
+When calling `getScope()` before a scope was made the current one (=before init was called), we will return a scope for a noop client. 
+A noop client is a regular client that simply does not send anything.
+
+This way, the API for `getClient()` can always return a client, and users do not have to guard against this being undefined all the time. 
+We may also expose a util like `sentryIsInitialized()` that checks if the current client is a Noop client (which currently you could have checked as `getCurrentHub().getClient() === undefined`).
+
+If you want to have multiple isolated clients, you can achieve this easily with this new setup:
+
+```js
+const client1 = new Client();
+const client2 = new Client();
+
+const scope1 = client1.getScope();
+const scope2 = client2.getScope();
+
+scope1.captureException(); // <-- isolated from scope2
+```
+
+## Scopes
+
+Scopes behave similar to how they behave today. 
+When a scope is forked via `withScope()`, a new scope is created that inherits all data currently set on the parent scope.
+
+The main change to Scopes is that they do not push/pop anymore, but instead fork an execution context (in languages where this makes sense/is possible).
+Basically, `withScope()` should behave like `hub.run()` does today.
+
+`client.getScope()` should return the current scope of this client in the current execution context.
+
+From a users perspective, this should mostly not be noticeable - they can always run `getScope()` to get the current scope, or `withScope(callback)` to fork a new scope off the current scope.
```

Comment:
This has been a major pain point in the past: `withScope()` was used in places where we actually wanted to change the current scope, but instead created a fork and then people were wondering why the changes were not visible on the original scope. Not sure if feasible within this RFC, but renaming `withScope()` to `forkScope()` would help here.

Reply:
I have no strong feelings about `withScope`, interested to hear what others think on this! I'd be fine with either.

Reply:
I find `forkScope` better, also considering this:

> You can still clone scopes manually the same way as before, e.g. via `Scope.clone(oldScope)` or a similar API. In contrast to `withScope()`, this will _not_ fork an execution context.


Both essentially create a new scope off of an existing one, but in different ways. It'd be good to have the name communicate what the difference is.

Reply:
Isn't `withScope()` mainly used to create a temporary scope which is only applied to the callback?
I don't see a good reason to rename/remove this.

Reply:
IMHO generally we should strive to replace this usage of `withScope` - only to add a temporary event processor - with different pattern(s). (not possible for _all_ cases, but most of them). So generally there should be little (not none) need to use it manually!